streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -518,168 +537,242 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
+            } else {
+              ZIO.succeedNow((s1, Some(chunk.drop(idx + 1))))
+            }
+        )
+      }
+
+    if (contFn(z))
+      ZSink[R, E, I, I, S] {
+        for {
+          state <- Ref.make(z).toManaged_
+          push = (is: Option[Chunk[I]]) =>
+            is match {
+              case None => state.get.flatMap(s => Push.emit(s, Chunk.empty))
+              case Some(is) => {
+                state.get.flatMap { s =>
+                  foldChunk(s, is, 0, is.length).foldM(err => Push.fail(err._1, err._2), {
+                    case (st, l) => {
+                      l match {
+                        case Some(leftover) => Push.emit(st, leftover)
+                        case None           => state.set(st) *> Push.more
+                      }
+                    }
+                  })
+                }
+              }
+            }
+        } yield push
+      }
+    else
+      ZSink.succeed(z)
+  }
 
   /**
    * A sink that folds its inputs with the provided function and initial state.
    */
-  def foldLeft[I, S](z: S)(f: (S, I) => S): ZSink[Any, Nothing, I, S] =
-    fold(z)(_ => true)(f)
+  def foldLeft[I, S](z: S)(f: (S, I) => S): ZSink[Any, Nothing, I, Nothing, S] =
+    fold(z)(_ => true)(f).dropLeftover
 
   /**
    * A sink that folds its input chunks with the provided function and initial state.
+   * `f` must preserve chunking-invariance.
    */
-  def foldLeftChunks[I, S](z: S)(f: (S, Chunk[I]) => S): ZSink[Any, Nothing, I, S] =
+  def foldLeftChunks[I, S](z: S)(f: (S, Chunk[I]) => S): ZSink[Any, Nothing, I, Nothing, S] =
     foldChunks(z)(_ => true)(f)
 
   /**
    * A sink that effectfully folds its input chunks with the provided function and initial state.
+   * `f` must preserve chunking-invariance.
    */
-  def foldLeftChunksM[R, E, I, S](z: S)(f: (S, Chunk[I]) => ZIO[R, E, S]): ZSink[R, E, I, S] =
-    foldChunksM[R, E, I, S](z: S)(_ => true)(f)
+  def foldLeftChunksM[R, E, I, S](z: S)(f: (S, Chunk[I]) => ZIO[R, E, S]): ZSink[R, E, I, Nothing, S] =
+    foldChunksM[R, E, I, S](z: S)(_ => true)(f).dropLeftover
 
   /**
    * A sink that effectfully folds its inputs with the provided function and initial state.
    */
-  def foldLeftM[R, E, I, S](z: S)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, S] =
+  def foldLeftM[R, E, I, S](z: S)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, I, S] =
     foldM[R, E, I, S](z: S)(_ => true)(f)
 
   /**
    * A sink that executes the provided effectful function for every element fed to it.
    */
-  def foreach[R, E, I](f: I => ZIO[R, E, Any]): ZSink[R, E, I, Unit] =
-    ZSink.fromPush {
-      case Some(is) => is.mapM_(f).mapError(Left(_))
-      case None     => ZIO.fail(Right(()))
+  def foreach[R, E, I](f: I => ZIO[R, E, Any]): ZSink[R, E, I, Nothing, Unit] =
+    ZSink.fromPush[R, E, I, Nothing, Unit] {
+      case Some(is) => is.mapM_(f).mapError(e => (Left(e), Chunk.empty))
+      case None     => Push.emit((), Chunk.empty)
     }
```

Comment:
If we have a `foldM` operation on ZSink, then this must preserve leftovers that occur when executing `mapM_` on the chunk.

E.g., `ZSink.fromEffect((i: Int) => if (i == 1) ZIO.fail("boom") else UIO.unit) orElse ZSink.fromEffect(i => UIO(println(i)))`, when fed `Chunk(1, 2)`, should not lose the `2`.

Reply:
Good catch, will add a test for it